octree_web_viewer/src/bin/points_web_viewer.rs

```diff
@@ -12,44 +12,121 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use octree_web_viewer::utils;
-use point_viewer::octree;
-use std::path::PathBuf;
+use octree_web_viewer::backend_error::PointsViewerError;
+use octree_web_viewer::state::AppState;
+use octree_web_viewer::utils::start_octree_server;
+use std::path::{Path, PathBuf};
 use std::sync::Arc;
+use structopt::clap::ArgGroup;
 use structopt::StructOpt;
 
+fn group_inputs() -> ArgGroup<'static> {
+    ArgGroup::with_name("path_stubs")
+        .conflicts_with("octree_path")
+        .requires_all(&["prefix", "suffix", "octree_id"])
+}
 /// HTTP web viewer for 3d points stored in OnDiskOctrees
 #[derive(StructOpt, Debug)]
-#[structopt(name = "points_web_viewer", about = "Visualizing points")]
-struct CommandlineArguments {
+#[structopt(
+    name = "points_web_viewer",
+    about = "Visualizing points",
+    raw(group = "group_inputs()")
+)]
+pub struct CommandLineArguments {
     /// The octree directory to serve, including a trailing slash.
-    octree_path: String,
+    /// this overrides the path_* options
+    #[structopt(name = "DIR", parse(from_os_str))]
+    octree_path: Option<PathBuf>,
     /// Port to listen on.
-    #[structopt(default_value = "5433", long = "--port")]
+    #[structopt(default_value = "5433", long = "port")]
     port: u16,
     /// IP string.
-    #[structopt(default_value = "127.0.0.1", long = "--ip")]
+    #[structopt(default_value = "127.0.0.1", long = "ip")]
     ip: String,
+    /// instead of DIR: specify path prefix for octree dir
+    #[structopt(long = "prefix", parse(from_os_str), group = "path_stubs")]
+    path_prefix: Option<PathBuf>,
+    /// Optional: suffix for subfolder of octree dir
+    #[structopt(long = "suffix", parse(from_os_str))]
+    path_suffix: Option<PathBuf>,
+    /// instead of DIR: specify path folder for octree dir
+    #[structopt(long = "octree_id", group = "path_stubs")]
+    octree_id: Option<String>,
+    /// Cache items
+    #[structopt(default_value = "20", long = "cache_items")]
+    cache_max: usize,
+}
+
+/// init app state with command arguments
+/// backward compatibilty is ensured
+pub fn state_from(args: CommandLineArguments) -> Result<AppState, PointsViewerError> {
+    let suffix = args.path_suffix.unwrap_or_else(|| PathBuf::from(""));
```

Comment:
Would `unwrap_or_default()` do the same thing?

Reply:
it won't compile